### PR TITLE
update block bloom properly

### PR DESCRIPTION
**Pull request by yihuang**
Closes: #139

Verified in external integration tests.

~~Depend on PR #192, since without that, the logs and bloom filter are always empty.~~

## Description

______

For contributor use:

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: x/evm/keeper/keeper.go [hunk 1]
```diff
@@ -117,24 +121,6 @@ func (k Keeper) SetBlockBloom(ctx sdk.Context, height int64, bloom ethtypes.Bloo
...
-	return new(big.Int).SetBytes(bz), true
-}
-
-// SetBlockBloomTransient sets the given bloom bytes to the transient store. This value is reset on
-// every block.
-func (k Keeper) SetBlockBloomTransient(bloom *big.Int) {
-	store := k.ctx.TransientStore(k.transientKey)
-	store.Set(types.KeyPrefixTransientBloom, bloom.Bytes())
-}
-
```
Comment (fedekunze) [inline x/evm/keeper/keeper.go:137]: why get rid of the transient store for the bloom filter?

File: x/evm/keeper/state_transition.go [hunk 2]
```diff
@@ -144,11 +144,16 @@ func (k *Keeper) ApplyTransaction(tx *ethtypes.Transaction) (*types.MsgEthereumT
...
+	// Remove once hashing is fixed on Tendermint. See https://github.com/tendermint/tendermint/issues/6539
+	txHash := common.BytesToHash(tmtypes.Tx(k.ctx.TxBytes()).Hash())
 	res.Hash = txHash.Hex()
-	res.Logs = types.NewLogsFromEth(k.GetTxLogs(txHash))
+
+	logs := k.GetTxLogs(txHash)
+	res.Logs = types.NewLogsFromEth(logs)
+
+	// update block bloom filter
+	k.Bloom.Or(k.Bloom, big.NewInt(0).SetBytes(ethtypes.LogsBloom(logs)))
```
Comment (fedekunze) [inline x/evm/keeper/state_transition.go:156]: set to transient store

File: x/evm/keeper/statedb.go [hunk 3]
```diff
@@ -591,7 +591,7 @@ func (k *Keeper) AddLog(log *ethtypes.Log) {
 	log.BlockHash = common.BytesToHash(k.ctx.HeaderHash())
 	log.TxIndex = uint(k.GetTxIndexTransient())
 
-	logs := k.GetTxLogs(log.TxHash)
+	logs := k.GetTxLogs(key)
```
Comment (fedekunze) [inline x/evm/keeper/statedb.go:594]: why this change?

**Review by fedekunze — CHANGES_REQUESTED** [review]
I feel the core issue here is that we were computing the book bloom differently. We should still use the transient store for keeping the bloom value as follows:

```go
k.SetBloom(ctx, bloom.Or(bloom, big.NewInt(0).SetBytes(ethtypes.LogsBloom(logs)))
```

Also, it would be nice to add additional tests for this logic

File: x/evm/keeper/statedb.go (hunk 3 above)
Comment (yihuang) [reply x/evm/keeper/statedb.go:594]: Because the `log.TxHash` is eth hash here, while `key` is the cosmos tx hash.

File: x/evm/keeper/keeper.go (hunk 1 above)
Comment (yihuang) [reply x/evm/keeper/keeper.go:137]: I'm not sure what's the difference between transient store and in memory directly.

Comment (yihuang) [reply x/evm/keeper/keeper.go]: done, changed to transient store.

File: x/evm/keeper/state_transition.go (hunk 2 above)
Comment (yihuang) [reply x/evm/keeper/state_transition.go]: done

**Comment by yihuang** [comment]
> I feel the core issue here is that we were computing the book bloom differently. We should still use the transient store for keeping the bloom value as follows:
> 
> ```go
> k.SetBloom(ctx, bloom.Or(bloom, big.NewInt(0).SetBytes(ethtypes.LogsBloom(logs)))
> ```

done

> 
> Also, it would be nice to add additional tests for this logic

to test this, I guess we have to call `BeginBlock`/`DeliverTx`/`EndBlock` in simulated env?

File: x/evm/keeper/keeper.go (hunk 1 above)
Comment (fedekunze) [reply x/evm/keeper/keeper.go]: Ref: https://docs.cosmos.network/master/core/store.html#transient-store

**Review by fedekunze — APPROVED** [review]
